### ISSUE TITLE
Files in .monacaignore aren't ignored on Reading Local Files step

### ISSUE THREAD

**Issue by corycaywood** [monaca/monaca-lib#125]
### Description
The `getLocalProjectFiles` method doesn't use the ignore patterns from the `.monacaignore` file when getting a listing of files in the project directory. This is a problem on projects with a large number of files in the `node_modules` folder because it will either take a very long time to get the file listing, or the JS heap will run out of memory.

This is the line that is causing the issue:
https://github.com/monaca/monaca-lib/blob/2.7.5/src/monaca.js#L1458

### Version
v2.7.5 is where the issue first occurs. 
The same issue does not seem to occur in v2.7.4 and earlier.

### Steps to Reproduce
1. Create a Cordova project with a large number of node module dependencies (~70 dependencies), and run `npm install`.
2. Create the following script in the root of your project:
```
#!/usr/bin/env node

const Monaca = require('monaca-lib').Monaca;
const monaca = new Monaca({clientType: 'cli'});

monaca.login("email@example.com", "password")
    .then(() => monaca.setProjectId(process.cwd(), "project_id"))
    .then(() => monaca.uploadProject(process.cwd()))
    .catch((error) => console.log(error));
```
3. Run the script: `node ./script.js`.

### Expected behaviour
Successfully read the local files and then uploads the files to your Monaca project.

### Actual behaviour
An Error is thrown:

```
Reading Local Files...
Reading Remote Files...

<--- Last few GCs --->

[27880:0x103000c00]    83925 ms: Mark-sweep 1400.4 (1511.0) -> 1400.4 (1480.0) MB, 2555.4 / 0.0 ms  (+ 0.0 ms in 0 steps since start of marking, biggest step 0.0 ms, walltime since start of marking 2555 ms) last resort
[27880:0x103000c00]    86320 ms: Mark-sweep 1400.4 (1480.0) -> 1400.4 (1480.0) MB, 2393.4 / 0.0 ms  last resort


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x1554ac228799 <JSObject>
    1: /* anonymous */(aka /* anonymous */) [PROJECT_PATH/node_modules/monaca-lib/node_modules/q/q.js:~1585] [pc=0xb56a773efb3](this=0x170969282311 <undefined>,undefined=0x170969282311<undefined>,promise=0x2d22d4abbf81 <Object map = 0x3cffc4b7ebc1>,index=17726)
    2: arguments adaptor frame: 4->3
    3: reduce(this=0x325b4d343c31 <JSArray[137441]>)
    5: /* an...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [~/.nvm/versions/node/v8.6.0/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [~/.nvm/versions/node/v8.6.0/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [~/.nvm/versions/node/v8.6.0/bin/node]
 4: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [~/.nvm/versions/node/v8.6.0/bin/node]
 5: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [~/.nvm/versions/node/v8.6.0/bin/node]
 6: 0xb56a75846fd
```

**Comment by masahirotanaka** [comment]
@yong-asial Could you please check this out? I think you're working on this fix now.

**Comment by yong-asial** [comment]
@corycaywood 
First of all, thank for reporting the issue. Unfortunately, I could not manage to reproduce the problem.
With the release of monaca-lib 2.7.5, one of the changes is that we will generate `.monacaignore` with predefine patterns. However, it won't generate if the file is already existed. Another changes made to monacaignore are (1) we remove the leading '/' and (2)  it now follows `.gitignore` patterns.

Previous `.monacaignore` patterns
```
/node_modules/**
/android/build/**
/ios/build/**
/build/**
*/.DS_Store
*/.git/**
```

Generated `.monacaignore` patterns
```
node_modules
.git
.monaca/*
!.monaca/project_info.json
android/build
ios/build
build
.DS_Store
```

For example, to completely ignore the `node_modules` directories (and its sub directories) we just have to write `node_modules` inside the `.monacaignore`.

Could you change the `.monacaignore` file and try again. Please let me know if the problem is still occurred.

P.S: in monaca-lib@2.7.4, `node_modules` is by default not included in the comparison logic regardless the `.monacaignore` content.

**Comment by corycaywood** [comment]
Hi @yong-asial! Thanks for the information about changes to the `.monacaignore` patterns. It's really nice that it's now using the same patterns as `.gitignore`.

Sorry, I think I made the title for this issue a little misleading. The files in `.monacaignore` are indeed being ignored by the file comparison logic and are not being uploaded to the Monaca project.

However, I think the problem is with this particular line, which uses `glob` to get an initial listing of all files in the project folder:
https://github.com/monaca/monaca-lib/blob/2.7.5/src/monaca.js#L1458

The glob pattern is `**/*`, which is fine for smaller projects, but for projects with many NPM dependencies it can cause the JavaScript heap to hit it's memory limit (at least on my machine). The `node_modules` folder for the project I'm testing on has about 140,000 files.

Even on smaller projects, this can make the upload step take significantly longer because there are so many files in `node_modules`. I have confirmed this by changing the glob pattern to `!(node_modules)**/*` and the upload step went much faster.

Thanks again! Let me know if you need any more information.

**Comment by yong-asial** [comment]
@corycaywood I see, I got your point now. 
I will work on it. Thank you for pointing out.

**Comment by yong-asial** [comment]
@corycaywood I've just released `monaca-lib@2.7.6`. I hope it should fix your problem. Please let us know if the issue is still existed.

**Comment by corycaywood** [comment]
@yong-asial : I tested v2.7.6 on my project with a large number of dependencies, and it worked correctly. The "Reading Local Files" step is also much faster now.

Thanks so much for your help!
